pkg/util/log: parse otlp sink from yaml config

TheComputerM opened the pull request:
OpenTelemetry is now an industry standard for o11y and is
more efficient than other log sinks currently available.

This commit only defines basic configuration options for
the OTLP sink, like address, insecure, and compression,
and adds logic to parse them from the YAML config. The
actual sink implementation will follow in a future commit.

Informs: https://github.com/cockroachdb/cockroach/issues/143049

Release note: None

arjunmahishi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147683)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, and @angles-n-daemons)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrRN45AhT6dnf4QHV2:-OSrRN45AhT6dnf4QHV3:b-b6nwo1):*
Add specific details about the changes in this commit. I.e It only adds the config parsing logic. the actual implementation will follow in a future commit.
___
*[`pkg/util/log/logconfig/config.go` line 606 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrRmaj-BdLxz-Ndgfh:-OSrRmaj-BdLxz-Ndgfi:undefined) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/config.go#L606)):*
> ```Go
> //
> // User-facing documentation follows.
> // TITLE: Output to OpenTelemetry collectors.
> ```





_Suggestion:_
```Go
Output to OpenTelemetry compatible collectors
```
___
*[`pkg/util/log/logconfig/config.go` line 634 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrS55q2uQbPvOXpQFo:-OSrS55q2uQbPvOXpQFp:undefined) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/config.go#L634)):*
> ```Go
> //	       # unless overridden here.
> //
> // The default output format for HTTP sinks is
> ```





_Suggestion:_
```Go
OTLP
```
___
*[`pkg/util/log/logconfig/config.go` line 640 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrOtPZ7qBrvGJC4c_1:-OSrOtPZ7qBrvGJC4c_2:bldbari) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/config.go#L640)):*
> ```Go
> // Run `cockroach debug check-log-config` to verify the effect of defaults inheritance.
> // {{site.data.alerts.end}}
> type OtlpSinkConfig struct {
> ```

The user will eventually be able to pick between GRPc and HTTP right?
I think we should have an option to configure that. We can keep "grpc" as default for now (since we don't have an HTTP implementation)
___
*[`pkg/util/log/logconfig/validate.go` line 523 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrPTdq1p0aBBj6UIby:-OSrPTdq1p0aBBj6UIbz:b-xy883e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/validate.go#L523)):*
> ```Go
> func (c *Config) validateOtlpSinkConfig(otsc *OtlpSinkConfig) error {
> 	propagateOtlpDefaults(&otsc.OtlpDefaults, c.OtlpDefaults)
> 	otsc.Address = strings.TrimSpace(otsc.Address)
> ```

We should not mutate a value in a "validate" function. For trimspace, can we find a more appropriate place? We might not have to do this at all actually. 

If this is only used for `len(otsc.Address) == 0` check, you can use TimeSpace inside the condition without actually mutating the value also

```
if len(strings.TimeSpace(otsc.Address)) == 0 {
```

---

Also, if the user has not specified the `Address`, we could consider defaulting to `localhost:standard_port`. What do you think?
___
*[`pkg/util/log/flags.go` line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrMlJS-L19PzC163w6:-OSrMlJS-L19PzC163w7:b-902nf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/flags.go#L450)):*
> ```Go
> 
> func newOtlpSinkInfo(_ logconfig.OtlpSinkConfig) (*sinkInfo, error) {
> 	return nil, nil
> ```

Why is this `nil`?
Is this left for a separate PR?


<!-- Sent from Reviewable.io -->

TheComputerM committed:
pkg/util/log: parse otlp sink from yaml config

OpenTelemetry is now an industry standard for o11y and is
more efficient than other log sinks currently available.

This commit only defines basic configuration options for
the OTLP sink, like address, insecure, and compression,
and adds logic to parse them from the YAML config. The
actual sink implementation will follow in a future commit.

Informs: https://github.com/cockroachdb/cockroach/issues/143049

Release note: None

TheComputerM reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147683)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrRN45AhT6dnf4QHV2:-OSrZ40WAM0kfZUa0DTP:b-896fix):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

Add specific details about the changes in this commit. I.e It only adds the config parsing logic. the actual implementation will follow in a future commit.
</blockquote></details>

Done.
___
*[`pkg/util/log/flags.go` line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrMlJS-L19PzC163w6:-OSrVvVl73M_TdMh_vGP:beulj0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/flags.go#L450)):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

Why is this `nil`?
Is this left for a separate PR?
</blockquote></details>

Yes, the actual sink functionality is left for a separate PR.
___
*[`pkg/util/log/logconfig/config.go` line 640 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrOtPZ7qBrvGJC4c_1:-OSrZ8BF14InI1dfmi5_:b-s4j3nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/config.go#L640)):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

The user will eventually be able to pick between GRPc and HTTP right?
I think we should have an option to configure that. We can keep "grpc" as default for now (since we don't have an HTTP implementation)
</blockquote></details>

I'll add the mode option, which will allow to toggle between gRPC and HTTP, in the same PR in which the HTTP mode will be implemented.
___
*[`pkg/util/log/logconfig/validate.go` line 523 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrPTdq1p0aBBj6UIby:-OSrWyeG9cCaoWt8fD6i:ba22hr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/validate.go#L523)):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

We should not mutate a value in a "validate" function. For trimspace, can we find a more appropriate place? We might not have to do this at all actually. 

If this is only used for `len(otsc.Address) == 0` check, you can use TimeSpace inside the condition without actually mutating the value also

```
if len(strings.TimeSpace(otsc.Address)) == 0 {
```

---

Also, if the user has not specified the `Address`, we could consider defaulting to `localhost:standard_port`. What do you think?
</blockquote></details>

The fluent sink does [the same](https://github.com/cockroachdb/cockroach/blob/28e1b226dfb3ceb38af3214d5b9e415d8e0c3c92/pkg/util/log/logconfig/validate.go#L447), I don't think its much of an issue as it just removes space from the ends of the string.

I don't think defaulting is a good idea in this case because if we choose to add HTTP support along with gRPC, the default address will change.
___
*[`pkg/util/log/logconfig/config.go` line 606 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrRmaj-BdLxz-Ndgfh:-OSrVaJEBpMEmv1d34UW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/config.go#L606)):*
> ```Go
> //
> // User-facing documentation follows.
> // TITLE: Output to OpenTelemetry collectors.
> ```

Done.
___
*[`pkg/util/log/logconfig/config.go` line 634 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrS55q2uQbPvOXpQFo:-OSrViVa3qkbRfVEdvb-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/logconfig/config.go#L634)):*
> ```Go
> //	       # unless overridden here.
> //
> // The default output format for HTTP sinks is
> ```

Done.


<!-- Sent from Reviewable.io -->

aa-joshi reviewed (commented):
> Reviewed 1 of 5 files at r1, 3 of 5 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147683)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`pkg/util/log/flags.go` line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrMlJS-L19PzC163w6:-OSwJJyKDkQwRfRAFDuC:bpfa8bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/flags.go#L450)):*
<details><summary><i>Previously, TheComputerM (Mudit Somani) wrote…</i></summary><blockquote>

Yes, the actual sink functionality is left for a separate PR.
</blockquote></details>

Can you add a TODO?
___
*[`pkg/util/log/logconfig/config.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSwJj7o8mojYBJWe-4U:-OSwJj7o8mojYBJWe-4V:bjst8q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/05280384997e209432b943dba4b1bbb897563a67/pkg/util/log/logconfig/config.go#L40)):*
> ```Go
> // DefaultOtlpFormat is the entry format for OpenTelemetry sinks
> // when not specified in a configuration.
> const DefaultOtlpFormat = `json`
> ```

qq: Are we planning to support any other format than json?
___
*[`docs/generated/logsinks.md` line 233 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSwJ6jbB55KWYpzL_te:-OSwJ6jbB55KWYpzL_tf:b-luadur) ([raw file](https://github.com/cockroachdb/cockroach/blob/05280384997e209432b943dba4b1bbb897563a67/docs/generated/logsinks.md#L233)):*
> ```Markdown
> 
> Configuration options shared across all sink types:
> ```

Is it auto generated?


<!-- Sent from Reviewable.io -->

TheComputerM reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147683)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`docs/generated/logsinks.md` line 233 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSwJ6jbB55KWYpzL_te:-OSwUp6y5qLsQBNFw7zz:b2nhswy) ([raw file](https://github.com/cockroachdb/cockroach/blob/05280384997e209432b943dba4b1bbb897563a67/docs/generated/logsinks.md#L233)):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Is it auto generated?
</blockquote></details>

Yes
___
*[`pkg/util/log/flags.go` line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSrMlJS-L19PzC163w6:-OSwVIAbCUSnZ9ji-aec:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f19ad21dfe2e1ab0cbd47d7f3fe63a6fd08c03f/pkg/util/log/flags.go#L450)):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Can you add a TODO?
</blockquote></details>

Done.
___
*[`pkg/util/log/logconfig/config.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSwJj7o8mojYBJWe-4U:-OSwUykSA3Q68A_5Q5-G:b-kyp7rk) ([raw file](https://github.com/cockroachdb/cockroach/blob/05280384997e209432b943dba4b1bbb897563a67/pkg/util/log/logconfig/config.go#L40)):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

qq: Are we planning to support any other format than json?
</blockquote></details>

Format is sink independent, we can send data in any format as long as the collector is configured to parse it.


<!-- Sent from Reviewable.io -->

arjunmahishi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Just one small NIT

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-:-OSx4DKF2vpzs2ktbLMh:bsjb7e3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, and @angles-n-daemons)

___
*[`-- commits` line 10 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSw_uVn8E7Rd5K-9EJs:-OSw_uVo4VVV2azCI1aV:b-41v31m):*
NIT 

Just suggesting a few semantic changes for more clarity



_Suggestion:_
```
  This commit only defines basic configuration options for the OTLP sink, like
  address, insecure, and compression, and adds logic to parse them from the YAML
  config. The actual sink implementation will follow in a future commit.
```


<!-- Sent from Reviewable.io -->

TheComputerM reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147683)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`-- commits` line 10 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSw_uVn8E7Rd5K-9EJs:-OSx6Y4bAelDfJt6J1su:b-896fix):*
<details><summary><i>Previously, arjunmahishi (Arjun Mahishi) wrote…</i></summary><blockquote>

Also, update the PR description with the same thing
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

TheComputerM commented:
bors r+

TheComputerM commented:
bors try

aa-joshi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-:-OT1N9UdBfUC_Gf9n-dW:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Abhinav1299, @angles-n-daemons, @arjunmahishi, and @TheComputerM)

___
*[`pkg/util/log/logconfig/config.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147683#-OSwJj7o8mojYBJWe-4U:-OT1N4CL8oMLEChkhIKA:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/05280384997e209432b943dba4b1bbb897563a67/pkg/util/log/logconfig/config.go#L40)):*
<details><summary><i>Previously, TheComputerM (Mudit Somani) wrote…</i></summary><blockquote>

Format is sink independent, we can send data in any format as long as the collector is configured to parse it.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->
